bundles/org.openhab.binding.velux/src/main/java/org/openhab/binding/velux/internal/bridge/slip/io/DataInputStreamWithTimeout.java

```diff
@@ -12,124 +12,199 @@
-     * @see java.io.DataInputStream#read
+     * @param timeoutMSecs the timeout period in milliseconds.
+     * @return the next SLIP message if there is one on the queue, or any empty byte[] array if not.
+     * @throws IOException
      */
-    public synchronized int read(byte b[], int off, int len, int timeoutMSecs) throws IOException {
-        // Definition of Method which encapsulates the Read of data
-        Callable<Integer> readTask = new Callable<Integer>() {
-            @Override
-            public Integer call() throws IOException {
-                return in.read(b, off, len);
+    public byte[] readSlipMessage(int timeoutMSecs) throws IOException {
+        startPolling();
+        int i = (timeoutMSecs / SLEEP_INTERVAL_MSECS) + 1;
+        while (i-- >= 0) {
+            try {
+                byte[] slip = slipMessageQueue.remove();
+                logger.trace("readSlipMessage() => return slip message");
+                return slip;
+            } catch (NoSuchElementException e) {
+                // queue empty, wait and continue
             }
-        };
-        try {
-            Future<Integer> future = executor.submit(readTask);
-            return future.get(timeoutMSecs, TimeUnit.MILLISECONDS);
-        } catch (RejectedExecutionException e) {
-            throw new IOException("executor failed", e);
-        } catch (ExecutionException e) {
-            throw new IOException("execution failed", e);
-        } catch (InterruptedException e) {
-            throw new IOException("read interrupted", e);
-        } catch (TimeoutException e) {
-            throw new IOException("read timeout", e);
+            try {
+                Thread.sleep(SLEEP_INTERVAL_MSECS);
+            } catch (InterruptedException e) {
+                logger.debug("readSlipMessage() => thread interrupt");
+                throw new IOException("Thread Interrupted");
+            }
+            throwIfPollException();
+        }
+        logger.debug("readSlipMessage() => no slip message after {}mS => time out", timeoutMSecs);
+        return new byte[0];
+    }
+
+    /**
+     * Get the number of incoming messages in the queue
+     *
+     * @return the number of incoming messages in the queue
+     */
+    public int available() {
+        int size = slipMessageQueue.size();
+        logger.trace("available() => slip message count {}", size);
+        return size;
+    }
+
+    /**
+     * Clear the queue
+     */
+    public void flush() {
+        logger.trace("flush() called");
+        slipMessageQueue.clear();
+    }
+
+    /**
+     * If necessary start the polling thread
+     */
+    private void startPolling() {
+        Thread pollThreadX = this.pollThread;
+        if (pollThreadX != null) {
+            logger.trace("startPolling()");
+            pollThreadX = pollThread = new Thread(pollRunner);
+            pollThreadX.setDaemon(true);
+            pollThreadX.setName(OH_BINDING_VELUX_BRIDGE);
```

Comment:
This should contain the Thing UID.

Reply:
I think these changes caused something to break in the binding. So I am testing again.

Reply:
@fwolter I think the problem is the `setDaemon(true)` call. Indeed from Googling an bit, I think there may be two issues with this..

- it seems to run at lower priority than a user thread; and I am certainly observing many read timeouts
- it does not shutdown cleanly; so when I drop a new jar build, it now requires me to reboot the hub

As it currently stands, it is not useable.

Reply:
https://www.baeldung.com/java-daemon-thread

Reply:
The cause of the read timeouts must be something different, as threads have the same priority as the thread which created them. See https://docs.oracle.com/javase/7/docs/api/java/lang/Thread.html second paragraph. Priority has nothing wether it's a daemon or not. The link you posted doesn't state anything contrary. It's only said that they are low-priority threads, but that's only a general statement.

Does the shutdown work if you disable the Thing in the UI?

Reply:
Yeah, it is probably the daemon bypassing of the polling thread's proper shutdown code.

The Velux KLF Bridge is very sensitive to being improperly shutdown. To do it properly, one must read all input notifications from its send buffer (i.e. clear the binding's data input stream), and then properly close the socket with a two way FIN/ACK. If one does not do that, then the KLF will go into a "zombie" state where it thinks it still has a socket open, but the binding has gone away. And the only way to eliminate the zombie state is to physically power cycle the unit. (Velux controls roof windows, and users do not appreciate having to crawl about in their attics to do that).

Now I have implemented all the socket communications as `Closeable` so that when one tries to shut down the binding, it will force all the sockets and streams to be properly finalized. Part of the orderly shutdown on the socket connection is the Close() action on the DataInputStream polling thread, to clear the Bridge's send buffer. And this works just fine if the DataInputStream polling thread is a user thread. However I suspect that if the DataInputStream polling thread is made into a daemon, then the binding unloading process kills the thread before the binding shutdown code's Close() gets called, and therefore it specifically leaves unread data on the the Bridge's send buffer, and therefore explicitly causes the KLF bridge to go into a zombie state. Requiring manual power cycling.

This can happen every time that the binding jar is un/re-reloaded and every time that OH is stopped/started. This is not good.

Reply:
If the KLF bridge is sensitive to improper shutdown then I think that use of a non-daemon thread is acceptable.

Reply:
OK